controllers/drain_controller_helper.go

```diff
@@ -0,0 +1,468 @@
+	if nodeStateDrainAnnotationCurrent == constants.DrainIdle {
+		result, err := dr.tryDrainNode(ctx, node)
+		if err != nil {
+			reqLogger.Error(err, "failed to check if we can drain the node")
+			return ctrl.Result{}, err
+		}
+
+		// in case we need to wait because we just to the max number of draining nodes
+		if result != nil {
+			return *result, nil
+		}
+	}
+
+	// call the drain function that will also call drain to other platform providers like openshift
+	drained, err := dr.drainer.DrainNode(ctx, node, nodeDrainAnnotation == constants.RebootRequired)
+	if err != nil {
+		reqLogger.Error(err, "error trying to drain the node")
+		dr.recorder.Event(nodeNetworkState,
+			corev1.EventTypeWarning,
+			"DrainController",
+			"failed to drain node")
+		return reconcile.Result{}, err
+	}
+
+	// if we didn't manage to complete the drain of the node we retry
+	if !drained {
+		reqLogger.Info("the nodes was not drained re queueing the request")
+		dr.recorder.Event(nodeNetworkState,
+			corev1.EventTypeWarning,
+			"DrainController",
+			"node drain operation was not completed")
+		return reconcile.Result{RequeueAfter: 5 * time.Second}, nil
+	}
+
+	reqLogger.Info("remove Device plugin from node")
+	err = utils.LabelNode(ctx, node.Name, constants.SriovDevicePluginEnabledLabel, constants.SriovDevicePluginEnabledLabelDisabled, dr.Client)
+	if err != nil {
+		reqLogger.Error(err, "failed to label node for device plugin label",
+			"labelKey",
+			constants.SriovDevicePluginEnabledLabel,
+			"labelValue",
+			constants.SriovDevicePluginEnabledLabelDisabled)
+		return reconcile.Result{}, err
+	}
+
+	// if we manage to drain we label the node state with drain completed and finish
+	err = utils.AnnotateObject(ctx, nodeNetworkState, constants.NodeStateDrainAnnotationCurrent, constants.DrainComplete, dr.Client)
+	if err != nil {
+		reqLogger.Error(err, "failed to annotate node with annotation", "annotation", constants.DrainComplete)
+		return ctrl.Result{}, err
+	}
+
+	reqLogger.Info("node drained successfully")
+	dr.recorder.Event(nodeNetworkState,
+		corev1.EventTypeWarning,
+		"DrainController",
+		"node drain completed")
+	return ctrl.Result{}, nil
+}
+
+func (dr *DrainReconcile) handleNodeDPReset(ctx context.Context,
+	reqLogger *logr.Logger,
+	node *corev1.Node,
+	nodeNetworkState *sriovnetworkv1.SriovNetworkNodeState,
+	nodeDrainAnnotation,
+	nodeStateDrainAnnotationCurrent string) (ctrl.Result, error) {
+	// nothing to do here we need to wait for the node to move back to idle
+	if nodeStateDrainAnnotationCurrent == constants.DrainComplete {
+		reqLogger.Info("node requested a drain and nodeState is on drain completed nothing todo")
+		return ctrl.Result{}, nil
+	}
+
+	// if we are on idle state we move it to drain
+	if nodeStateDrainAnnotationCurrent == constants.DrainIdle {
+		err := utils.AnnotateObject(ctx, nodeNetworkState, constants.NodeStateDrainAnnotationCurrent, constants.Draining, dr.Client)
```

Comment:
It took me a while to understand this. We're setting `nodeState = Draining` even if we are not draining. We're just cordoning. It is to avoid having too many states?

Reply:
right I didn't want to make this more complex than it's already :P